sync/class.jetpack-sync-module-import.php

```diff
@@ -1,87 +1,97 @@
 <?php
 
+require_once dirname( __FILE__ ) . '/class.jetpack-sync-functions.php';
+
 class Jetpack_Sync_Module_Import extends Jetpack_Sync_Module {
 
-	private $import_end = false;
+	/**
+	 * Tracks which actions have already been synced for the import
+	 * to prevent the same event from being triggered a second time
+	 * @var array
+	 */
+	private $synced_actions = array();
+
+	/**
+	 * A mapping of known importers to friendly names.
+	 * Keys are the class name of the known importer.
+	 * Values are the friendly name.
+	 * @var array
+	 */
+	private static $known_importers = array(
+		'Blogger_Importer' => 'blogger',
+		'LJ_API_Import' => 'livejournal',
+		'MT_Import' => 'mt',
+		'RSS_Import' => 'rss',
+		'WC_Tax_Rate_Importer' => 'woo-tax-rate',
+		'WP_Import' => 'wordpress',
+	);
+
+	/**
+	 * A mapping of action types to sync action name.
+	 * Keys are the name of the import action.
+	 * Values are the resulting sync action.
+	 * @var array
+	 */
+	private static $import_sync_action_map = array(
+		'import_start' => 'jetpack_sync_import_start',
+		'import_done'  => 'jetpack_sync_import_end',
+		'import_end'   => 'jetpack_sync_import_end',
+	);
 
 	public function name() {
 		return 'import';
 	}
 
 	public function init_listeners( $callable ) {
 		add_action( 'export_wp', $callable );
-		add_action( 'jetpack_sync_import_end', $callable, 10, 2 );
+		add_action( 'jetpack_sync_import_start', $callable, 10, 2 );
+		add_action( 'jetpack_sync_import_end',   $callable, 10, 2 );
+
+		// WordPress
+		add_action( 'import_start', array( $this, 'sync_import_action' ) );
 
 		// Movable type, RSS, Livejournal
-		add_action( 'import_done', array( $this, 'sync_import_done' ) );
+		add_action( 'import_done',  array( $this, 'sync_import_action' ) );
 
 		// WordPress, Blogger, Livejournal, woo tax rate
-		add_action( 'import_end', array( $this, 'sync_import_end' ) );
+		add_action( 'import_end',   array( $this, 'sync_import_action' ) );
 	}
 
 	public function set_defaults() {
-		$this->import_end = false;
+		$this->synced_actions = array();
 	}
 
-	public function sync_import_done( $importer ) {
-		// We already ran an send the import
-		if ( $this->import_end ) {
-			return;
-		}
+	public function sync_import_action( $importer ) {
+		$import_action = current_filter();
+		// map action to event name
+		$sync_action = self::$import_sync_action_map[ $import_action ];
 
-		$importer_name = $this->get_importer_name( $importer );
-
-		/**
-		 * Sync Event that tells that the import is finished
-		 *
-		 * @since 5.0.0
-		 *
-		 * $param string $importer
-		 */
-		do_action( 'jetpack_sync_import_end', $importer, $importer_name );
-		$this->import_end = true;
-	}
-
-	public function sync_import_end() {
-		// We already ran an send the import
-		if ( $this->import_end ) {
+		// Only sync each action once per import
+		if ( array_key_exists( $sync_action, $this->synced_actions ) && $this->synced_actions[ $sync_action ] ) {
```

Comment:
If some request triggers both `import_done` & `import_end` it looks like this would limit it to a single `jetpack_sync_import_end`, correct?  That seems like the desired behavior here, so 👍 if so :)

Is there any instance where these would be triggered in separate request cycles for the same import?  If so, is it acceptable that we'd have repeat sync events?  Probably not a huge issue if so.

Reply:
That is correct. The replaced code worked similarly (though, it only handled a single action mapping, ensuring `jetpac_sync_import_end` was only triggered once). The update 

While I suppose that it's possible there is an importer that would trigger multiple `import_done` or `import_end` actions outside of a single request, none of the primary importers that we're syncing currently do. If we find this to be the case, we can certainly revisit -- but since it's something that importers _shouldn't_ do, until and unless we see it in practice, I think it's acceptable.